src/app/modules/payments/stripe.service.ts

```diff
@@ -1009,15 +1011,155 @@ export const verifyPaymentStatusWithStripe = (
     })
 }
 
-export const generatePaymentInvoice = (
+const getPaymentProofObjectPath = (payment: IPaymentSchema) => {
+  return payment.formId + '/' + payment._id + '.pdf'
+}
+
+/**
+ * Function that stores payment proof into s3
+ *
+ * @param {IPaymentSchema} payment the payment object, used to form the object path
+ * @param {Buffer} pdfBuffer the pdf to store into s3
+ *
+ * @returns ok(undefined) if no errors are thrown while uploading to s3
+ * @returns err(InvoiceUploadS3Error) if an error is thrown while uploading to s3
+ */
+const _storePaymentProofInS3 = (
+  payment: IPaymentSchema,
+  pdfbuffer: Buffer,
+): ResultAsync<undefined, PaymentProofUploadS3Error> => {
+  const objectPath = getPaymentProofObjectPath(payment)
+
+  logger.info({
+    message: 'Uploading payment proof to s3',
+    meta: {
+      action: '_storePaymentProofInS3',
+      paymentId: payment._id,
+      objectPath,
+      bucket: AwsConfig.paymentProofS3Bucket,
+    },
+  })
+
+  return ResultAsync.fromPromise(
+    AwsConfig.s3
+      .upload({
+        Bucket: AwsConfig.paymentProofS3Bucket,
+        Key: objectPath,
+        Body: Buffer.from(pdfbuffer),
+      })
+      .promise(),
+    (error) => {
+      logger.error({
+        message: 'Error occured whilst uploading pdfBuffer to S3',
+        meta: {
+          action: 'storePaymentProofInS3',
+          paymentId: payment._id,
+          objectPath,
+        },
+        error,
+      })
+      return new PaymentProofUploadS3Error()
+    },
+  )
+    .map(() => {
+      if (payment.completedPayment) {
```

Comment:
qn (probably because of lack of context) : why is the completedPayment check here, should this be at the start of the fn i.e. if the payment hasn't been completed, we should not be storing payment proof in s3?

Reply:
Mainly due to lack of refinement of `payment.completePayment`, it was ready checked in the previous chain under `_retrieveReceiptUrlFromStripe`.

Added typing refinement in 18c84b271